src/vpn-api/app.ts

```diff
@@ -35,34 +39,103 @@ import worker from './worker';
 
 const VPN_INSTANCE_COUNT =
 	parseInt(process.env.VPN_INSTANCE_COUNT!, 10) || os.cpus().length;
+const VPN_BYTECOUNT_INTERVAL =
+	parseInt(process.env.VPN_BYTECOUNT_INTERVAL!, 10) || 0;
 
 if (cluster.isMaster) {
 	logger.info(
 		`open-balena-vpn@${VERSION} master process started with pid ${process.pid}`,
 	);
-	if (VPN_INSTANCE_COUNT > 1) {
-		logger.info(`spawning ${VPN_INSTANCE_COUNT} workers`);
-		_.times(VPN_INSTANCE_COUNT, i => {
-			const instanceId = i + 1;
-			const restartWorker = (code?: number, signal?: string) => {
-				if (signal != null) {
-					logger.error(
-						`open-balena-vpn worker-${instanceId} killed with signal ${signal}`,
-					);
-				}
-				if (code != null) {
-					logger.error(
-						`open-balena-vpn worker-${instanceId} exited with code ${code}`,
-					);
-				}
-				cluster.fork({ VPN_INSTANCE_ID: instanceId }).on('exit', restartWorker);
-			};
-			restartWorker();
+	logger.info(`spawning ${VPN_INSTANCE_COUNT} workers`);
+	_.times(VPN_INSTANCE_COUNT, i => {
+		const instanceId = i + 1;
+		const restartWorker = (code?: number, signal?: string) => {
+			if (signal != null) {
+				logger.error(
+					`open-balena-vpn worker-${instanceId} killed with signal ${signal}`,
+				);
+			}
+			if (code != null) {
+				logger.error(
+					`open-balena-vpn worker-${instanceId} exited with code ${code}`,
+				);
+			}
+			cluster.fork({ VPN_INSTANCE_ID: instanceId }).on('exit', restartWorker);
+		};
+		restartWorker();
+	});
+
+	// track session stats from workers
+	const sessions: {
+		[key: string]: {
+			ts: number;
+			rxBytes: number;
+			txBytes: number;
+		};
+	} = {};
+
+	if (VPN_BYTECOUNT_INTERVAL > 0) {
+		cluster.on('message', (_worker, msg) => {
+			const { uuid, ts, rxBytes, txBytes } = msg;
+			sessions[uuid] = { ts, rxBytes, txBytes };
 		});
 	}
+
+	const kb = 2 ** 10; // 1024
+	const mb = 2 ** 10 * kb;
+	const gb = 2 ** 10 * mb;
+	const tb = 2 ** 10 * gb;
+	const buckets = [
+		kb,
+		mb,
+		10 * mb,
+		100 * mb,
+		500 * mb,
+		gb,
+		10 * gb,
+		100 * gb,
+		150 * gb,
+		250 * gb,
+		500 * gb,
+		tb,
+	];
+
+	// setup metrics for prometheus
+	metrics.describe(
+		'vpn_sessions_rx_bytes',
+		'histogram of rx bytes per vpn session',
+		{
+			buckets,
+			labelNames: [],
+		},
+	);
+	metrics.describe(
+		'vpn_sessions_tx_bytes',
+		'histogram of tx bytes per vpn session',
+		{
+			buckets,
+			labelNames: [],
+		},
+	);
+
+	const app = express();
+	app.disable('x-powered-by');
+	app.get('/ping', (_req, res) => res.send('OK'));
+	app.use(morgan('combined'));
+	app.get('/metrics', (_req, res) => {
```

Comment:
There's actually, in `@balena/node-metrics-gatherer` @ `v2.2.1`, a better pattern for aggregating metrics of `cluster` workers, which uses a pattern designed by the prometheus client library for this purpose. See [the changed `index.js` in `resin-api` for an example](https://github.com/balena-io/resin-api/blob/b95ad206402a009c80b3902f829ec3dcc9d19fc0/index.js).

In this case, you'd do something like:

- replace `process.send` in the `cluster` worker code with a simple call to metrics, like `metrics.histogram('vpn_session_rx_bytes', rxBytes)`

- in the master thread, do:

```
app.get('/cluster_metrics', metrics.aggregateRequestHandler());
``` 

(replaceing the entire block from L125 to L134)

Reply:
That is to say, the workers perform metrics calls as if they weren't in a cluster, while the master, in the internals of `prom-client` which `metrics.aggregateRequestHandler()` wraps, will manage the IPC and summation logic which gathers each cluster workers' metrics when the time comes to export.

Reply:
Oh sweet, thanks for the heads up!